.github/workflows/linter.yml

```diff
@@ -0,0 +1,59 @@
+###########################
+###########################
+## Linter GitHub Actions ##
+###########################
+###########################
+name: Lint Code Base
+
+#
+# Documentation:
+# https://help.github.com/en/articles/workflow-syntax-for-github-actions
+#
+# For the Web Almanac we have the following rule changes from the defaults:
+#  * Markdown - we're a bit more relaxed about this since authors may be less technical
+#  * Python - Only change from default is to set ignored-modules=flask, flask-talisman...etc.
+#  * CSS - Tweaked to ignore some Prince/Normalise exceptions and some whitespace exceptions
+#  * JS - default rules
+#  * JSON - default rules
+#  * YML - default rules
+# Exception config files are in the .github/linters directory
+
+######################################
+# Start the job on all pull requests #
+######################################
+on: pull_request
+
+###############
+# Set the Job #
+###############
+jobs:
+  build:
```

Comment:
Linting only on PRs means people will not get their code linted as they commit into their forks before creating a PR.

```suggestion
#########################
# Lint each push and PR #
#########################
on:
  - push
  - pull_request

###############
# Set the Job #
###############
jobs:
  lint:
    ########################################
    # Skip duplicate job for local brances #
    ########################################
    if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
```

Reply:
That’s true. Though on other hand some might get annoyed at running a whole lint job on every commit if code is work in progress. But guess they can just ignore the feedback.

Can you explain that last line to me? So that means it will run on all pushes, but only in pull requests from forks?

If so I’ve two questions:
1. We probably don’t need to run this on pushes to `main` in this primary repo, after accepting a PR. But then do want to run on `main` for forks as some people do develop on that.
2. Does this mean it will not run on ably PRs from this repo? Some of us work in branches in the main repo rather than forks and I would like to know when the PR is submitted that the linters are all green.

Reply:
It is very important to run lint and test tasks on each push to ensure no push is causing any new issues. These CI processes run asynchronously, so they should not bother anyone. On the other hand, one can leverage them to incrementally improve the work in progress. Note that these jobs will only run when one or more commits are actually pushed to GH infrastructure, not when we are committing changes incrementally on our local development machines.

It is also important to run these tasks on PRs to ensure that the maintainers decide what failures they are willing to tolerate, not the contributor who may have created a PR with failing tests/lints in her/his fork.

This means, when maintainers create a branch in this repo (not in a separate fork) and then create a PR to the main branch, these jobs will be triggered twice at the same time in the same repo, one for the pus action and the other for the PR. This is where the `if:` condition comes in handy. It will let all the pushes run the job, but in case of PRs, it will only run name of the head repo is not the same as this repo (i.e., skipping jobs on PRs from the local branches). I will note here that in case of skipped jobs, we will see an ugly looking skipped task in the PR, but that is better than simultaneous duplicate run. I hope GH comes up with a better solution for this scenario.

Finally, we can skip this job to run when changes are pushed (or PRs are merged) into teh main branch, but there is actually no harm if we let it run once again asynchronously. That's why I did not exclude the main branch. This also means, if someone with write permission accidentally pushes something to the main branch directly, at least those changes will be linted/tested.

Reply:
I have a slightly different take:

It is very important for them to run on PRs, as that means the results can be checked before the PR is merged. The result is then also shown in the pull result itself allowing a reviewer to check that as part of their PR review.

It would be very handy to run them on every commit, so developers can correct as they go along, and won't have a load of corrections to do at the end, after they've opened the pull request. The downside to this, is that there is no GUI to note the fact it fails (like in pull requests). Yes the pusher should get an email, and it's available in the Actions tab but not quite as nice as seeing a nicely linked status in the pull request. Another downside is GitHub actions can take time, and you are limited to the amount you can run. So two or three commits in quick succession could create a queue (how often have we all hit "push" – or "send" on an email – and then immediately spotted a typo!). The fact it only runs on the changed files is good, but the action still takes a few mins, to launch, checkout that branch, download the action docker file...etc.

So I agreed it would be ideal to have both, but to me doing it on a PR is a _must-have_ for me to protect the code, and doing it on a push is a _nice-to-have_.

My concern with not doing it on PRs for branches with-in this repo is would the results show in the pull request? Let's say I opened a branch (with-in this repo), then committed bad code to that branch, ignored the lint failures, and then opened a pull request. I thought with your change would me the pull request reviewer would not see the results of the action in the pull request. But testing it, it looks like that is not the case - it _does_ show the action of the last action in the pull request. Excellent - didn't realise it did that.

So have made your change and will merge if you are happy now?

Reply:
> My concern with not doing it on PRs for branches with-in this repo is would the results show in the pull request?

Yes, two tasks will show up in the PR that will be triggered due to two two different events (push and PR). However, they will essentially be doing the same thing simultaneously, so it is better to skip one of them. That's why I suggested the scary-looking `if:` condition.